NDTensors/src/lib/UnallocatedArrays/src/unallocatedfill.jl

```diff
@@ -0,0 +1,87 @@
+using FillArrays: FillArrays, AbstractFill, Fill, broadcasted_fill, kron_fill, mult_fill
+using NDTensors.SetParameters: Position, set_parameters
+import Base: +
+
+struct UnallocatedFill{ElT,N,Axes,Alloc} <: AbstractFill{ElT,N,Axes}
+  f::Fill{ElT,N,Axes}
+  alloc::Alloc
+
+  function UnallocatedFill{ElT,N,Axes}(
+    f::Fill, alloc::Type{<:AbstractArray{ElT,N}}
+  ) where {ElT,N,Axes}
+    return new{ElT,N,Axes,Type{alloc}}(f, alloc)
+  end
+end
+
+function UnallocatedFill{ElT,N}(f::Fill, alloc) where {ElT,N}
+  return UnallocatedFill{ElT,N,typeof(axes(f))}(f, alloc)
+end
+
+function UnallocatedFill{ElT}(f::Fill, alloc) where {ElT}
+  return UnallocatedFill{ElT,ndims(f)}(f, alloc)
+end
+
+function UnallocatedFill(f::Fill, alloc)
+  return UnallocatedFill{eltype(f)}(f, alloc)
+end
+
+set_alloctype(f::Fill, alloc::Type{<:AbstractArray}) = UnallocatedFill(f, alloc)
+
+Base.parent(F::UnallocatedFill) = F.f
+
+Base.convert(::Type{<:UnallocatedFill}, A::UnallocatedFill) = A
+
+#############################################
+# Arithmatic
+
+# mult_fill(a, b, val, ax) = Fill(val, ax)
+function FillArrays.mult_fill(a::UnallocatedFill, b, val, ax)
+  return UnallocatedFill(Fill(val, ax), alloctype(a))
+end
+FillArrays.mult_fill(a, b::UnallocatedFill, val, ax) = mult_fill(b, a, val, ax)
+function FillArrays.mult_fill(a::UnallocatedFill, b::UnallocatedFill, val, ax)
+  @assert(alloctype(a) == alloctype(b))
+  return UnallocatedFill(Fill(val, ax), alloctype(a))
+end
+
+function FillArrays.broadcasted_fill(f, a::UnallocatedFill, val, ax)
+  return UnallocatedFill(Fill(val, ax), alloctype(a))
```

Comment:
An issue I see here is that `val` and `ax` may change the element type and number of dimensions, so the `alloc` field might not be correct any more (i.e. inconsistent with the new `Fill`).

For now we could just check for that and throw an error and fix it later. Should be easy enough to adjust `alloctype(a)` using `set_eltype` and `set_ndims`.

Another possibility is to try to think of a deeper solution for this kind of issue, since I'm sure it would come up a lot. I could imagine two `UnallocatedFill` constructors, one that takes the `alloc` input literally, and one that adjusts it based on the properties of the `Fill` input.

Also a tricky part is dealing with other objects that could be in the `alloc` field, like the compute units in `HeterogenousComputing.jl`. Might be tricky to generically adjust the element type and number of dimensions of those.

In the end, the simpler solution could be to just be less strict about the details of the `alloc` field, and then define `allocate(f::UnallocatedFill)` could make use of `alloctype(f)` and also `eltype(parent(f))` and `axes(parent(f))`, for example something like:
```julia
function allocate(f::UnallocatedFill)
  a = similar(alloctype(f), eltype(f), axes(f))
  fill!(a, getindex_value(f))
  return a
end
```
Then the element type and number of dimensions of the `alloc` field are effectively ignored.

Reply:
Or:
```julia
function allocate(f::UnallocatedFill)
  a = similar(f)
  fill!(a, getindex_value(f))
  return a
end
```
and define:
```julia
function Base.similar(f::UnallocatedFill)
  return similar(alloctype(f), eltype(f), axes(f))
end
```
So the only requirement of `alloctype(f)` is that given an element type and axes it allocates an array when `similar` is called.

Reply:
@kmp5VT what do you think of this suggestion? By redefining `allocate` and `similar` in this way we can be more relaxed about the type of the `alloc` field and we could simplify some of the code in this PR, i.e. it would circumvent some uses of `set_parameters`/`set_eltype` on `alloctype(f)` since we wouldn't have to be so careful about adjusting the type.

Reply:
@mtfishman I am looking more into the FillArrays and it looks like they use the `convert` function to allocate.  So based on their convention our allocate could look like 
```
allocate(A::UnallocatedFill) = convert(alloctype(A), parent(A))
```
Would this be sufficient or I can also define allocate and similar based on your suggestions!

Reply:
The issue with that design is that if `alloc` is set to the wrong element type (i.e. one inconsistent with the element type of the parent Fill array), it would give the wrong result.

Reply:
Right now that problem is mitigated in the constructor of the `UnallocatedArray` because I only have a single constructor which expects alloc to have the same eltype as the AbstractFIll
```
function UnallocatedFill{ElT,N,Axes}(
    f::Fill, alloc::Type{<:AbstractArray{ElT,N}}
  ) where {ElT,N,Axes}
    return new{ElT,N,Axes,Type{alloc}}(f, alloc)
  end
  ``` 
  but I understand this won't be the case if we use the Heterogenous computing units. 
  
  I wrote `allocate` using `similar` like you suggested however it doesn't look there exists a similar which takes eltype and a Tuple.
  ```
  similar(Matrix, Float32, axes(Z))
ERROR: MethodError: no method matching similar(::Type{Matrix}, ::Type{Float32}, ::Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}})
```

Reply:
Good point about `similar`, I could have sworn it worked on types but I guess not. As an alternative we could do something like this:
```julia
function Base.similar(f::UnallocatedFill)
  return allocate(alloctype(f), eltype(f), axes(f))
end

allocate(arraytype::Type{<:AbstractArray}, elt::Type, axes::Tuple) = set_eltype(arraytype, elt)(undef, axes)
```

Reply:
Okay I wrote something similar. One thing is that I believe `set_ndims` and `set_eltypes` are functions which are defined by NDTensors, not SetParameters. So instead I used 
```
function allocate(arraytype::Type{<:AbstractArray}, elt::Type, axes::Tuple)
  ArrayT = set_parameters(arraytype, Position{1}(), elt)
  ArrayT = set_parameters(ArrayT, Position{2}(), length(axes))
  return similar(ArrayT, axes)
end
```
But potentially these are functions which can be put in SetParameters and defined for different datatypes.

Reply:
Hmm that's too bad that we can't use `set_eltype` and `set_ndims` here because of that. Could you leave a comment about that? We should move them to `SetParameters` in a future PR.

Reply:
Okay I just made a comment about this in the code and wrote a section in the `TODO.md` of `SetParameters`